Update README for continuous aggregates (#2405)
timescale/timescaledb: tsl/src/continuous_aggs/Readme.md

last hour's worth of data) or ensure that the continuous aggregate is
always up-to-date with the underlying source data. Policies that focus
on recent data allow older parts of the continuous aggregate to stay
the same or be governed by manual refreshes.

## Bookkeeping and Internal State ##

TimescaleDB does bookkeeping for each continuous aggregate to know
which buckets of the aggregates require refreshing. Whenever a
modification happens to the source data, an invalidation for the
modified region is written to an invalidation log. However,
invalidations are not written beyond the *invalidation threshold*,
which tracks the largest bucket materialized thus far. This threshold
allows write amplification to be kept to a minimum by, e.g, never
refreshing the most recent time bucket that is still being "filled".

Thus, to store, maintain, and query aggregations, continuous
aggregates consists for the following objects and state:

1. A query view, which finalizes the aggregations and is also the
   object that users interact with.
2. A partial view, which is used to materialize new data.
3. A direct view, which holds the original query that users specified.

> 4. An internal materialization hypertable, containing the materialized
>    data for each time bucket.

Reviewer: ```suggestion
4. An internal materialization hypertable, containing the materialized
   data as partial aggregates for each time bucket.
```

5. An invalidation threshold, which tracks how "far" the
   materialization has reached and below which invalidations must be
   logged.
6. An invalidation log. Entries in this log contain time ranges that
   should be re-materialized during refreshes, as an INSERT, UPDATE,
   DELETE, or other mutating operation may have altered the source
   data. In practice, this log is split in two; once a refresh is run
   on a continuous aggregate, each new invalidation is copied to a
   per-materialization log that tracks which invalidations have been
   processed for each continuous aggregate. This is necessary since
   there might be multiple continuous aggregate per source hypertable,

>    and each of them might be refreshed and materialized separately.

Reviewer: Maybe describe the two logs as two separate objects in that case. I think it would also make the description easier to follow.

rather the partial aggregate state. For instance, in case of an
average, each bucket stores the sum and count in an internal binary
form. The partial aggregates are what gives continuous aggregates
flexibility; buckets can be individually updated and multiple partial
aggregates can be combined to form new partials. Future enhancements
may allow aggregating at different time resolutions using the the same
underlying continuous aggregate.

## The Invalidation Log and Threshold ##

Mutating transactions must record their mutations in the invalidation
log, so that a refresh knows to re-materialize the invalidated
range. This happens by installing a trigger on the source hypertable
when the first continuous aggregate on that hypertable is created.

To reduce the extra writes by the trigger, only one invalidation range
(lowest and highest modified value) is written at the end of a
mutating transaction. As a result, a refresh might materialize more
data than necessary, but the insert incurs a smaller overhead
instead. Write amplification is further reduced by never writing
invalidations that modify time ranges that are more recent the current

> invalidation threshold.

Reviewer: It is not clear why this reduces write amplification.

buckets). Finalization of the partial buckets happens automatically at
query time. Although such finalization gives slightly higher querying
times, it is offset by more efficient refreshes that only recompute
the buckets that have been "invalidated" by changes in the raw data.

A continuous aggregate policy automates the refreshing, allowing the
aggregate to stay up-to-date without manual intervention. A policy can
be configured to only refresh the most recent data (e.g., just the
last hour's worth of data) or ensure that the continuous aggregate is
always up-to-date with the underlying source data. Policies that focus
on recent data allow older parts of the continuous aggregate to stay
the same or be governed by manual refreshes.

## Bookkeeping and Internal State ##

TimescaleDB does bookkeeping for each continuous aggregate to know
which buckets of the aggregates require refreshing. Whenever a
modification happens to the source data, an invalidation for the
modified region is written to an invalidation log. However,
invalidations are not written after the *invalidation threshold*,
which tracks the latest bucket materialized thus far. This threshold

> allows write amplification to be kept to a minimum by, e.g, never
> refreshing the most recent time bucket that is still being "filled".

Reviewer: Are there any other things we currently do? Otherwise, just write what we do.
```suggestion
allows write amplification to be kept to a minimum by not
refreshing the most recent time bucket, which is assumed to still have data being added to it.
```

Thus, to store, maintain, and query aggregations, continuous
aggregates consist of the following objects:

1. A user view, which queries and finalizes the aggregations and is
   also the object that users interact with.
2. A partial view, which is used to materialize new data.
3. A direct view, which holds the original query that users specified.
4. An internal materialization hypertable, containing the materialized
   data as partial aggregates for each time bucket.

> 5. An invalidation threshold, which tracks how "far" the
>    materialization has reached and below which invalidations must be
>    logged.

Reviewer: ```suggestion
5. An invalidation threshold, which is a timestamp that tracks the latest 
   materialization. Invalidations before this timestamp will be
   logged, but invalidations after this point will not be added to the invalidation log.
```

> 6. A trigger on the source hypertable that invalidates regions of data
>    in every transaction, based on INSERT, UPDATE, and DELETE
>    statements. The trigger writes its invalidations to the hypertable
>    invalidation log.

Reviewer: It doesn't actually invalidate the regions, it just writes an invalidation record to the log. IMHO, it is better to be specific here.
```suggestion
6. A trigger on the source hypertable that writes invalidations of data
   for every transaction, based on INSERT, UPDATE, and DELETE
   statements. The trigger writes its invalidations to the hypertable
   invalidation log.
```

7. A hypertable invalidation log that tracks invalidated regions of
   data for each hypertable. Entries in this log contain time ranges
   that need to be re-materialized across all the hypertable's
   continuous aggregates.

> 8. A materialization invalidation log. Once a refresh runs on a given
>    continuous aggregate, this log tracks how invalidations from the
>    hypertable invalidation log are processed against the refresh
>    window for the refreshed continuous aggregate. Thus, a single
>    invalidation in the hypertable invalidation log becomes one entry
>    per continuous aggregate in the materialization invalidation log.

Reviewer: ```suggestion
8. A materialization invalidation log for each continuous aggregate. Once a refresh runs on a given
   continuous aggregate, this log tracks how invalidations from the
   hypertable invalidation log are processed against the refresh
   window for the refreshed continuous aggregate. Thus, a single
   invalidation in the hypertable invalidation log becomes one entry
   per continuous aggregate in the materialization invalidation log.
```

Author: There is only one log, and your suggestion could be interpreted as if there is one log per cagg.

## The materialized hypertable ##

The materialized hypertable does not store the aggregate's output, but
rather the partial aggregate state. For instance, in case of an
average, each bucket stores the sum and count in an internal binary
form. The partial aggregates are what gives continuous aggregates
flexibility; buckets can be individually updated and multiple partial
aggregates can be combined to form new partials. Future enhancements
may allow aggregating at different time resolutions using the the same
underlying continuous aggregate.

## The Invalidation Log and Threshold ##

> Mutating transactions must record their mutations in the invalidation
> log, so that a refresh knows to re-materialize the invalidated
> range. This happens by installing a trigger on the source hypertable
> when the first continuous aggregate on that hypertable is created.

Reviewer: ```suggestion
Mutating transactions must record their mutations in the invalidation
log for the hypertable, so that a refresh knows to re-materialize the invalidated
range. This happens by installing a trigger on the source hypertable
when the first continuous aggregate on that hypertable is created.
```

Author: Again, your suggestion could be interpreted as if there is one log per hypertable, which is not the case.
